File: 430-flatten-a-multilevel-doubly-linked-list.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
}
*/

public class Solution {
    public Node Flatten(Node head) {
        if (head == null) return head;

        var dummy = new Node(0, null, head, null);
        Node curr, prev = dummy;

        var stack = new Stack<Node>();
        stack.Push(head);

        while (stack.Count != 0) {
            curr = stack.Pop();
            prev.next = curr;
            curr.prev = prev;

            if (curr.next != null) stack.Push(curr.next);
            if (curr.child != null) {
                stack.Push(curr.child);
                curr.child = null;
            }

            prev = curr;
        }

        dummy.next.prev = null;
        return dummy.next;
    }
}

File: 430-flatten-a-multilevel-doubly-linked-list.go
package main

type Node struct {
	Val   int
	Prev  *Node
	Next  *Node
	Child *Node
}

func flatten(root *Node) *Node {
	if root == nil {
		return root
	}

	stack := []*Node{root}
	dummy := &Node{0, nil, root, nil}
	prev := dummy
	for len(stack) != 0 {
		cur := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		prev.Next = cur
		cur.Prev = prev
		if cur.Next != nil {
			stack = append(stack, cur.Next)
		}
		if cur.Child != nil {
			stack = append(stack, cur.Child)
			cur.Child = nil
		}

		prev = cur
	}

	dummy.Next.Prev = nil
	return dummy.Next
}
